docs/src/generative.md

```diff
@@ -1,49 +1,85 @@
+## Random number generators
 
-Use the following logic to determine which functions to implement:
-- If you are implementing the problem from scratch in Julia, implement `generate_s` and `generate_o`.
-- Otherwise, if your external simulator returns *x*, where *x* is one of *sr*, *so*, *or*, or *sor*, implement `generate_x`. (you may also have to implement `generate_s` separately for use in particle filters).
+The `rng` argument to functions in the generative interface is a random number generator such as `Base.GLOBAL_RNG` or another `MersenneTwister`. It should be used to generate all random numbers within the function (e.g. use `rand(rng)` instead of `rand()`). This will ensure that all simulations are exactly repeatable. See the [Julia documentation on random numbers](https://docs.julialang.org/en/v1/stdlib/Random/#Random-Numbers-1) for more information about these objects.
 
-Note: if an explicit definition is already implemented, you **do not** need to implement any functions from the generative interface - POMDPs.jl will automatically generate implementations of them for you at runtime (see generative_impl.jl).
+## Performance considerations
 
-### Solver and Simulator Writers
+In general, calling `gen(::DBNOut, ...)` when `gen(::POMDP, ...)` is implemented does not introduce much overhead. In fact, in some cases, the compiler will even optimize out calculations of extra genvars. For example:
+```julia
+struct M <: MDP{Int, Int} end
+
+POMDPs.gen(::M, s, a, rng) = (sp=s+a, r=s^2)
 
-Solver writers should use the single function that generates everything that they need and nothing they don't. For example, if the solver needs access to the state, observation, and reward at every timestep, they should use `generate_sor()` rather than `generate_s()` and `generate_or()`, and if the solver needs access to the state and reward, they should use `generate_sr()` rather than `generate_sor()`. This will ensure the widest interoperability between solvers and problems.
+@code_warntype gen(Return(:sp), M(), 1, 1, Random.GLOBAL_RNG)
+```
+will yield
+```
+Body::Int64
+1 ─ %1 = (Base.add_int)(s, a)::Int64
+│        nothing
+└──      return %1
+```
+indicating that the compiler will only perform the addition to find the next state and skip the `s^2` calculation for the reward.
 
-In other words, if you need access to *x* where *x* is *s*, *o*, *sr*, *so*, *or*, or *sor* at a certain point in your code, use `generate_x`.
+Unfortunately, if random numbers are used, the compiler will not be able to optimize out the change in the rng's state, so it may be beneficial to directly implement versions of `gen` with specified return values.
```

Comment:
The subsequent example makes this clear, but it might be worth changing to "if random numbers are used in the `gen(::POMDP,...)` method..."

Reply:
done